src/hotspot/share/runtime/atomic.hpp

```diff
@@ -576,6 +690,99 @@ struct Atomic::PlatformXchg {
                atomic_memory_order order) const;
 };
 
+// Implement fetch_then_bitop operations using a CAS loop.
+class Atomic::PrefetchBitopsUsingCmpxchg {
+  template<typename T, typename Op>
+  T bitop(T volatile* dest, atomic_memory_order order, Op operation) const {
+    T old_value;
+    T new_value;
+    T fetched_value = Atomic::load(dest);
+    do {
+      old_value = fetched_value;
+      new_value = operation(old_value);
+      fetched_value = Atomic::cmpxchg(dest, old_value, new_value, order);
```

Comment:
At some point in future, it would be nice to replace the retry-loops like these with _weak_ CASes. It could be done in platform-specific code, though, but it would seem awkward to specialize for all LL/SC-retry-loop platforms just to get access to weak CAS.

Reply:
Yes.  I haven't checked, but I would guess the majority of our CAS uses are loops that would benefit from
using a weak CAS on platforms where that's available.